docs/content/guides/developer/app-examples/reviews-rating.mdx

```diff
@@ -0,0 +1,212 @@
+
+The reviews are rated (scored) on-chain using following criteria
+- Intrinsic score (IS)
+  - Length of content
+- Extrinsic score (ES)
+  - Number of votes received
+- Verification multiplier (VM)
+  - Reviews with PoE get a multiplier to improve rating
+
+```
+Total Score = (IS + ES) * VM
+```
+
+## Smart contracts
+
+There are several modules that create the backend logic for the example. 
+
+### dashboard.move
+
+The `dashboard.move` module defines the `Dashboard` struct that is used to group services.
+
+```move
+struct Dashboard has key, store {
+  id: UID,
+  set: VecSet<ID>,
+  service_type: String
+}
+```
+
+The services are grouped by attributes, which can be cuisine type, geographical location, operating hours, google maps ID, and so on. To keep it basic, the example stores only `service_type` (for example, fast food, Chinese, Italian).
+
+### service.move
+
+This module defines the `Services` struct, which service owners manage.
+
+```move
+struct Service has key, store {
+  id: UID,
+  reward_pool: Balance<SUI>,
+  reward: u64,
+  reviews: MultiMap<ID>,
+  moderators: VecSet<address>,
+  overall_rate: u64,
+  name: String
+}
+```
+
+All the submitted reviews are stored in `reviews` field, which has `MultiMap<ID>` type. The [multimap module](#multimap) defines the type. 
+
+The `reward_pool` and `reward` fields store the available reward information.  
+
+```move
+struct AdminCap has key, store {
+  id: UID,
+  service_id: ID
+}
+struct Moderator has key {
+  id: UID,
+  service_id: ID,
+}
+```
+`SERVICE OWNERS` are given `AdminCap` and `MODERATORS` are given `Moderator` to allow them to perform previleged operations.
+
+### multimap.move {#multimap}
+
+`MultiMap` is an implementation of the `multimap` data structure (https://en.wikipedia.org/wiki/Multimap), which keeps all the reviews sorted. Using `MultiMap`, all the reviews are ordered by rating, with the highest rated reviews coming first.
+
+```move
+struct MultiMap<K: copy> has copy, drop, store {
+  contents: vector<Entry<K>>,
+}
+struct Entry<K: copy> has copy, drop, store {
+  key: K,
+  priority: u64,
+}
+```
+The `object_id` of a `review` object is used as a `key` and `total_score` is used as `priority` for a new entry.
+
+Note that `vector` is used to store the contents of the `MultiMap`. This is acceptable in our use case because the maximum number of reviews for a service is 1000.
+As reviews are sorted by `total_score` on-chain, it is necessary to keep a manageable number of reviews for a service.
+
+```move
+public fun insert<K: copy>(self: &mut MultiMap<K>, key: K, priority: u64) {
+  assert!(!contains(self, &key), EKeyAlreadyExists);
+  let len = vector::length(&self.contents);
+  if (len == 0) {
+    vector::push_back(&mut self.contents, Entry { key, priority });
+  } else {
+    insert_recursive(&mut self.contents, Entry { key, priority }, 0, len - 1);
+  }
+}
+
+fun insert_recursive<K: copy>(v: &mut vector<Entry<K>>, e: Entry<K>, left: u64, right: u64) {
```

Comment:
Best not to implement this recursively -- function calls are expensive in Move and max stack size is limited.  Here is a non-recursive implementation of binary search:

https://github.com/MystenLabs/sui/blob/f3a9d828880b80617d517416e52eccdfdd715401/examples/move/big_vector/sources/big_vector.move#L576-L593

Reply:
Used non-recursive function for binary search as per suggestion.